tests/integration/test_xrdsst.py

```diff
@@ -0,0 +1,127 @@
+                  'configuration_anchor': configuration_anchor,
+                  'owner_dn_country': 'EE',
+                  'owner_dn_org': 'ORG',
+                  'owner_member_class': 'GOV',
+                  'owner_member_code': '1234',
+                  'security_server_code': 'SS',
+                  'software_token_id': 0,
+                  'software_token_pin': '1234'}]}
+        return config
+
+    def setUp(self):
+        config = self.single_ss_config()
+        name = config["security-server"][0]["name"]
+        if not os.path.exists(self.local_folder):
+            git.Repo.clone_from(self.git_repo, self.local_folder, branch=self.branch_name)
+        client = docker.from_env()
+        images = client.images.list(name=self.image)
+        if len(images) == 0:
+            client.images.build(path=self.docker_folder, tag=self.image)
+        containers = client.containers.list(all=True, filters={"name": name})
+        if len(containers) == 0:
+            client.containers.run(detach=True, name=name, image=self.image, ports={'4000/tcp': 8000})
+        containers = client.containers.list(all=True, filters={"name": name})
+        for container in containers:
+            if container.name == name:
+                retries = 0
+                while retries <= self.max_retries:
+                    cmd = "curl -X POST -u xrd:secret --silent " + self.url + " --data \'" + self.roles + \
+                          "\'" + " --header \"" + self.header + "\" -k"
+                    result = container.exec_run(cmd, stdout=True, stderr=True, demux=False)
+                    if len(result.output) > 0:
+                        json_data = json.loads(result.output)
+                        self.api_key = json_data["key"]
+                        break
+                    time.sleep(self.curl_retry_wait_seconds)
+                    retries += 1
+
+    def tearDown(self):
+        subprocess.call("rm -rf " + self.local_folder + "/", shell=True)
+        subprocess.call("docker ps -a | awk '{ print $1,$2 }' | grep " + self.image +
+                        " | awk '{print $1 }' | xargs -I {} docker rm -f {}", shell=True)
+        subprocess.call("docker rmi -f $(docker images --format '{{.Repository}}:{{.Tag}}' "
+                        "| grep \'" + self.image + "\')", shell=True)
+
+    def step_init(self):
+        config = self.single_ss_config()
+        config["security-server"][0]["api_key"] = 'X-Road-apikey token=' + self.api_key
+        base = BaseController()
+        init = InitServerController()
+        configuration = base.initialize_basic_config_values(config["security-server"][0])
+        status = init.check_init_status(configuration)
+        assert status.is_anchor_imported is False
+        assert status.is_server_code_initialized is False
+        init.initialize_server(config)
+        status = init.check_init_status(configuration)
+        assert status.is_anchor_imported is True
```

Comment:
Local run after successful acquisition of API key failed to import anchor due to ``conf_verification.unreachable``:

```
==================================================================================================== FAILURES =====================================================================================================
________________________________________________________________________________________ TestXRDSST.test_run_configuration ________________________________________________________________________________________

self = <tests.integration.test_xrdsst.TestXRDSST testMethod=test_run_configuration>

    def test_run_configuration(self):
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
>       self.step_init()

tests/integration/test_xrdsst.py:124: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.integration.test_xrdsst.TestXRDSST testMethod=test_run_configuration>

    def step_init(self):
        config = self.single_ss_config()
        config["security-server"][0]["api_key"] = 'X-Road-apikey token=' + self.api_key
        base = BaseController()
        init = InitServerController()
        configuration = base.initialize_basic_config_values(config["security-server"][0])
        status = init.check_init_status(configuration)
        assert status.is_anchor_imported is False
        assert status.is_server_code_initialized is False
        init.initialize_server(config)
        status = init.check_init_status(configuration)
>       assert status.is_anchor_imported is True
E       AssertionError: assert False is True
E        +  where False = {'is_anchor_imported': False,\n 'is_server_code_initialized': False,\n 'is_server_owner_initialized': False,\n 'software_token_init_status': 'NOT_INITIALIZED'}.is_anchor_imported

tests/integration/test_xrdsst.py:96: AssertionError
---------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------
Starting configuration process for security server: ss
Uploading configuration anchor for security server: ss
Exception when calling SystemApi->upload_initial_anchor: (500)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '26a8521d27e9eb9c', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Thu, 10 Dec 2020 15:09:25 GMT', 'Connection': 'close'})
HTTP response body: {"status":500,"error":{"code":"conf_verification.unreachable"}}


Initializing security server: ss
Exception when calling InitializationApi->init_security_server: (409)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '283f8bb5a6961e29', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Thu, 10 Dec 2020 15:09:25 GMT'})
HTTP response body: {"status":409,"error":{"code":"anchor_not_found"}}


db1bc23a51fe
Untagged: xroad-security-server:latest
Deleted: sha256:684eac8a5b3877e3a584e190d7d3a0cf3d6c0b2a5843ceda1f9c3b53e96b06bd
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
ERROR    root:init.py:77 Exception when calling SystemApi->upload_initial_anchor: (500)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '26a8521d27e9eb9c', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Thu, 10 Dec 2020 15:09:25 GMT', 'Connection': 'close'})
HTTP response body: {"status":500,"error":{"code":"conf_verification.unreachable"}}


ERROR    root:init.py:97 Exception when calling InitializationApi->init_security_server: (409)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': '283f8bb5a6961e29', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Thu, 10 Dec 2020 15:09:25 GMT'})
HTTP response body: {"status":409,"error":{"code":"anchor_not_found"}}

----------- coverage: platform linux, python 3.8.5-final-0 -----------
Name                              Stmts   Miss  Cover
-----------------------------------------------------
xrdsst/controllers/base.py           49      5    90%
xrdsst/controllers/init.py           72      3    96%
xrdsst/controllers/timestamp.py      87      9    90%
xrdsst/controllers/token.py         148     14    91%
-----------------------------------------------------
TOTAL                               356     31    91%
Coverage HTML written to dir coverage-report

============================================================================================= short test summary info =============================================================================================
FAILED tests/integration/test_xrdsst.py::TestXRDSST::test_run_configuration - AssertionError: assert False is True
==================================================================================== 1 failed, 45 passed in 371.44s (0:06:11) =====================================================================================
make: *** [Makefile:34: test-all] Error 1
```

Reply:
I have updated the code quite a bit, please test again. In addition, now the configuration anchor is taken as parameter:
    configuration_anchor = "tests/resources/configuration-anchor.xml"

But in any case, I think it is best to keep the anchor in "test/resources" folder, you should locally just have to copy that over with your own configuration anchor. The same way, a script will copy the configuration anchor on the Jenkins server from "/etc/xroad" to that folder. That is sort of expected and can not be avoided in any way, that you need to use your own anchor and the tests that we are going to run on Jenkins, will be using an anchor from a central X-Road CS according to Ilkka, so we just need to copy the file to the test resources.